File: mycobot_ros/mycobot_320/new_mycobot_320_moveit/scripts/agv_socket.py
#!/usr/bin/env python
import rospy
from std_msgs.msg import String
import socket
import time

Avalue_received = False
Bvalue_received = False

# ROS 노드 초기화
rospy.init_node('step_publisher_node', anonymous=True)

def send_value_callback(msg):
    global Avalue_received, Bvalue_received

    if msg.data == "A":
        print("sending A signal to socket")
        if not Avalue_received:
            Avalue_received = True
    
    if msg.data == "B":
        print("sending B signal to socket")
        if not Bvalue_received:
            Bvalue_received = True

# Subscriber 초기화
rospy.Subscriber("done_value", String, send_value_callback)

#소켓 초기화
client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
server_address = ('172.30.1.29', 8080)
print("Connecting to server at", server_address)
client_socket.connect(server_address)

def keyboard_listener():
    global Avalue_received, Bvalue_received

    key = input("Press 'A' or 'B': ")
    if key == 'A':
        Avalue_received = True
        print("Avalue_received set to True")
        main()

    elif key == 'B':
        Bvalue_received = True
        print("Bvalue_received set to True")
        main()    

def main():
    global Avalue_received, Bvalue_received

    while not rospy.is_shutdown():
        try:
            if Bvalue_received:
                Bvalue_received = False
                # "B" 메시지 송신
                try:
                    message = "3"
                    print("Sending:", message)
                    client_socket.sendall(message.encode())
                    time.sleep(5)
                    keyboard_listener()
                except Exception as e:
                    print("Error occurred:", str(e))
            if Avalue_received:
                Avalue_received = False
                # "A" 메시지 송신
                try:
                    message = "4"
                    print("Sending:", message)
                    client_socket.sendall(message.encode())
                    time.sleep(5)
                    keyboard_listener()
                except Exception as e:
                    print("Error occurred:", str(e))

        except Exception as e:
            print("Error occurred:", str(e))

if __name__ == '__main__':
    # main()  # 메인 루프 실행
    keyboard_listener()

File: project3/scripts/agv_socket.cpp
#include <ros/ros.h>
#include <std_msgs/String.h>
#include <iostream>
#include <string>
#include <boost/asio.hpp>

bool Avalue_received = false;
bool Bvalue_received = false;

using namespace std;
using namespace ros;

boost::asio::io_service io_service;
boost::asio::ip::tcp::socket socket(io_service);

void send_value_callback(const std_msgs::String::ConstPtr& msg) {
    if (msg->data == "A") {
        cout << "sending A signal to socket" << endl;
        if (!Avalue_received) {
            Avalue_received = true;
        }
    }

    if (msg->data == "B") {
        cout << "sending B signal to socket" << endl;
        if (!Bvalue_received) {
            Bvalue_received = true;
        }
    }
}

void keyboard_listener() {
    cout << "Press 'A' or 'B': ";
    char key;
    cin >> key;

    if (key == 'A') {
        Avalue_received = true;
        cout << "Avalue_received set to True" << endl;
    } else if (key == 'B') {
        Bvalue_received = true;
        cout << "Bvalue_received set to True" << endl;
    }
}

void main_loop() {
    while (ros::ok()) {
        try {
            if (Bvalue_received) {
                Bvalue_received = false;
                string message = "1";
                cout << "Sending: " << message << endl;
                boost::asio::write(socket, boost::asio::buffer(message));
                Duration(5).sleep();
                keyboard_listener();
            }
            if (Avalue_received) {
                Avalue_received = false;
                string message = "2";
                cout << "Sending: " << message << endl;
                boost::asio::write(socket, boost::asio::buffer(message));
                Duration(5).sleep();
                keyboard_listener();
            }
        } catch (exception& e) {
            cerr << "Error occurred: " << e.what() << endl;
        }
    }
}

int main(int argc, char** argv) {
    init(argc, argv, "step_publisher_node");
    NodeHandle nh;

    Subscriber sub = nh.subscribe("done_value", 1000, send_value_callback);

    boost::asio::ip::tcp::endpoint endpoint(boost::asio::ip::address::from_string("172.30.1.29"), 8080);
    socket.connect(endpoint);

    keyboard_listener();
    main_loop();

    return 0;
}
